src/persist-client/src/write.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::marker::PhantomData;
+
+use differential_dataflow::difference::Semigroup;
+use mz_persist_types::{Codec, Codec64};
+use timely::progress::{Antichain, Timestamp};
+use tracing::warn;
+
+use crate::error::Error;
+
+pub struct WriteHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> WriteHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Semigroup + Codec64,
+{
+    // This handle's upper frontier, not the global collection-level one.
+    pub fn upper(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
+    // that no more data is ever incoming
+    //
+    // TODO: Bound the write and space amplification as a function of how much
+    // data has even been written to this collection as well as how much of it
+    // is "live" (doesn't consolidate out at the collection frontier).
+    pub async fn write_batch<I: IntoIterator<Item = ((K, V), T, D)>>(
```

Comment:
Not to get into the weeds here, but it may make sense to have these be references since I assume they will be encoded into a byte buffer and not used directly. Correct me if I'm wrong

Reply:
no you're absolutely right. we have a couple things like this now and so far have just ended up using the owned version so the reference one was a usability papercut, but I agree that if we're only going to have one, it should be the one that makes better perf possible. is this what you're thinking?

> <I: IntoIterator<Item = ((&'_ K, &'_ V), &'_ T, &'_ D)>>

if not, just type what you want and we can make it work